pkg/ingester/rate_calculator.go

```diff
@@ -0,0 +1,61 @@
+package ingester
+
+import (
+	"sync"
+	"time"
+
+	"go.uber.org/atomic"
+)
+
+// Samples are recorded once every 1000 / bucketCount milliseconds.
+const bucketCount = 100
+
+// RateCalculator buckets rate measurements and returns a per-second rate of the recorded values
+type RateCalculator struct {
+	lock    sync.RWMutex
+	buckets []int64
+	idx     int
+	sample  atomic.Int64
+}
+
+func NewRateCalculator() *RateCalculator {
+	c := &RateCalculator{
+		buckets: make([]int64, bucketCount),
+	}
+
+	go c.recordSample()
+
+	return c
+}
+
+func (c *RateCalculator) recordSample() {
+	t := time.NewTicker(1000 / bucketCount * time.Millisecond)
+	defer t.Stop()
```

Comment:
This isn't going to do anything, unfortunately. We need to add a `Stop` method to the rate limiter than can be called when the stream is removed otherwise this goroutine will hang around forever.